### ISSUE TITLE
Type predicate assertion does not carry over when used in another function

### ISSUE THREAD

**Issue by yogiduzit** [microsoft/TypeScript#51326]
# Bug Report

### 🔎 Search Terms
- predicate
- type predicate
- type predicate return

### 🕗 Version & Regression Information

- This is NOT a crash
- This changed between versions ______ and _______
- This is the behavior in every version I tried, and I reviewed the FAQ for entries about _________
- I was unable to test this on prior versions because _______

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.6.4#code/JYOwLgpgTgZghgYwgAgIImAWzgG2QbwChlkQ5MIAuZAZzClAHMBuQgX0MNElkRQCFgUACbIIAD0ghhNNBmx4iJAO5MAygAc4IanQYgW7TgHoAVKeKm0NGtDCywACzhhk2t-NzJgsgPYxkMABPDRQAFW8AhF9pYDBgGO8HKABXFBhfKEDnV20sXAA6S2NCGBSQBHjEn3R8nAB5GDCQiAAeCIkpGTk6gD4ACjyFalqFABpkaNiqnWRBzxxqMIBKZABeXuQAI19fHAhtZeohrx9kCKVkKAgwFKgQSZjhOISQebq3WRXWDlLyyteSUEInewx6ClWl2ut3uSVGuEazVCrWBwgGJxwE36WyEwlWG2Q-F2+202NxBVUBk0h2WP04ZQqM2QjBuAHUmDQtG8cSJqKjIcQrjc7g8ecIKeouXTCFM6B46rI1sgANr4UjkKjIABEcC1yDYEzVZAo1B1WomlMYNFNAGYYGA9WwALqfcG4ZVO1h-RmAhCOCAIADW8JwNFBuGtbpwHoFJGhIvlChoBWwGnDeAJlxIsr2EAKOF8jHTtMFJGAAX6PlRxdjJDrOf2+cL-RZYHZBk5pIxyxLdf1grYyyMnCAA)

### 💻 Code

```ts
interface Animal {
  name: string;
}

interface Bird extends Animal {
  wingSpan: string;
}

/**
 * Asserts that an animal is of type T if condition is true for that animal.
 */
function isAnimalOfType<T extends Animal>(animal: Animal, condition: (animal: T) => boolean): animal is T {
  return condition(animal as T);
}

function isBird(animal: Animal) {
  return isAnimalOfType<Bird>(animal, (bird) => Boolean(bird.wingSpan));
}

function getWingspan(bird: Bird) {
  return bird.wingSpan;
}

const animals = [{ name: "a" }, { name: "a", wings: "3ft" }] as Animal[];

function checkAnimals(animals: Animal[]) {
  return animals.map(animal => {
    console.log(animal);
    if (isBird(animal)) {
      console.log(getWingspan(animal));
    }
  })
}
```

### 🙁 Actual behavior

The call to `getWingspan` shows an error saying `Argument of type 'Animal' is not assignable to parameter of type 'Bird'`.

### 🙂 Expected behavior

The function `isAnimalOfType` already asserts that the Animal will be of type T. So, why does `isBird` not automatically assert that?

**Comment by fatcerberus** [comment]
Type predicates are not automatically inferred.  You need to explicitly declare `isBird` as returning `animal is Bird`.

**Comment by jcalz** [comment]
Duplicate of [#10734](https://github.com/microsoft/TypeScript/issues/10734#issuecomment-245134817)

**Comment by danvk** [comment]
This was marked as a duplicate of #10734, which was closed by #57465. But that PR only let "identifier" type predicates flow, not assertion predicates. The example in this issue is still broken.

This specific issue _would_ be fixed by #58495, see [playground].

[playground]: https://www.typescriptlang.org/play/?ts=5.5.0-pr-58495-20#code/JYOwLgpgTgZghgYwgAgIImAWzgG2QbwChlkQ5MIAuZAZzClAHMBuQgX0MNElkRQCFgUACbIIAD0ghhNNBmx4iJAO5MAygAc4IanQYgW7TgHoAVKeKm0NGtDCywACzhhk2t-NzJgsgPYxkMABPDRQAFW8AhF9pYDBgGO8HKABXFBhfKEDnV20sXAA6S2NCGBSQBHjEn3R8nAB5GDCQiAAeCIkpGTk6gD4ACjyFalqFABpkaNiqnWRBzxxqMIBKZABeXuQAI19fHAhtZeohrx9kCKVkKAgwFKgQSZjhOISQebq3WRXWDlLyyteSUEInewx6ClWl2ut3uSVGuEazVCrWBwgGJxwE36WyEwlWG2Q-F2+202NxBVUBk0h2WP04ZQqM2QjBuAHUmDQtG8cSJqKjIcQrjc7g8ecIKeouXTCFM6B46rI1sgANr4UjkKjIABEcC1yDYEzVZAo1B1WomlMYNFNAGYYGA9WwALqfcG4ZVO1h-RmAhCOCAIADW8JwNFBuGtbpwHoFJGhIvlChoBWwGnDeAJlxIsr2EAKOF8jHTtMFJGAAX6PlRxdjJDrOf2+cL-RZYHZBk5pIxyxLdf1grYyyMnCAA
